### PR TITLE
[Template] use generated outputs as hash source

### PR DESCRIPTION
**Pull request by BumbleFeng**
restructure TemplateInput to be TemplateResult

### CODE REVIEW

File: reconcile/templating/renderer.py [hunk 1]
```diff
@@ -269,9 +269,9 @@ def reconcile_template_collection(
         gql_api: GqlApi,
         persistence: FilePersistence,
         ruamel_instance: yaml.YAML,
-        input: TemplateInput,
+        result: TemplateResult,
         each: dict[str, Any],
-    ) -> list[TemplateOutput]:
+    ) -> None:
```
Comment (hemslo) [inline reconcile/templating/renderer.py:274]: the old signature is easier to understand, can we remove the `result` from params, and still return `list[TemplateOutput]`?

File: reconcile/templating/lib/model.py [hunk 2]
```diff
@@ -1,34 +1,23 @@
...
-class TemplateOutput(BaseModel):
-    input: TemplateInput
-    is_new: bool = False
-    path: str
-    content: str
-    auto_approved: bool = False
+        if not self.result_hash:
+            hashable = {"outputs": sorted(self.outputs, key=lambda x: x.path)}
+            self.result_hash = DeepHash(hashable)[hashable]
+        return self.result_hash
```
Comment (hemslo) [inline reconcile/templating/lib/model.py:23]: This method is to calc hash only once, right? Can use [cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property) to simplify it, example to use with `BaseModel` at https://github.com/app-sre/qontract-reconcile/blob/1b80a586850a1145aff072f7837b71bb716d223d/reconcile/utils/early_exit_cache.py#L76.

To make hash safe and stable, should also mark `TemplateOutput` as frozen. If `path` is unique, can make `hashable` a dict `{o.path: o for o in self.outputs}`

Comment (BumbleFeng) [reply reconcile/templating/lib/model.py]: I found out calc hash would only run once when submitting MR, so there is no need to add cache.

File: reconcile/templating/lib/merge_request_manager.py [hunk 3]
```diff
@@ -121,34 +121,26 @@ def process(self, gitlab_cli: GitLabApi) -> None:
...
-
-
 class MergeRequestManager(MergeRequestManagerBase[TemplateInfo]):
     def __init__(self, vcs: VCS, parser: Parser):
         super().__init__(vcs, parser, TR_LABEL)
 
-    def create_merge_request(self, data: MrData) -> None:
+    def create_merge_request(
+        self, data: TemplateResult, auto_approved: bool = False
+    ) -> None:
```
Comment (hemslo) [inline reconcile/templating/lib/merge_request_manager.py:130]: in `MergeRequestManagerBase`, `create_merge_request` is an abstract method, better to keep params (`data`) unchanged.

**Review by hemslo — APPROVED** [review]
LGTM!

**Pull request merged**